Modules/posixmodule.c

```diff
@@ -7379,6 +7378,86 @@ os_openpty_impl(PyObject *module)
 #endif /* defined(HAVE_OPENPTY) || defined(HAVE__GETPTY) || defined(HAVE_DEV_PTMX) */
 
 
+#ifdef HAVE_SETSID
+#if defined(TIOCSCTTY) || defined(HAVE_TTYNAME)
+#define HAVE_FALLBACK_LOGIN_TTY 1
+#endif /* defined(TIOCSCTTY) || defined(HAVE_TTYNAME) */
+#endif /* HAVE_SETSID */
+
+#if defined(HAVE_LOGIN_TTY) || defined(HAVE_FALLBACK_LOGIN_TTY)
+/*[clinic input]
+os.login_tty
+
+    fd: fildes
+    /
+
+Prepare the tty of which fd is a file descriptor for a new login session.
+
+Make the calling process a session leader; make the tty the
+controlling tty, the stdin, the stdout, and the stderr of the
+calling process; close fd.
+[clinic start generated code]*/
+
+static PyObject *
+os_login_tty_impl(PyObject *module, int fd)
+/*[clinic end generated code: output=495a79911b4cc1bc input=5f298565099903a2]*/
+{
+#ifdef HAVE_LOGIN_TTY
+    if (login_tty(fd) == -1) {
+        return posix_error();
+    }
+#else /* defined(HAVE_FALLBACK_LOGIN_TTY) */
+    /* Establish a new session. */
+    if (setsid() == -1) {
+        return posix_error();
+    }
+
+    /* The tty becomes the controlling terminal. */
+#ifdef TIOCSCTTY
+    if (ioctl(fd, TIOCSCTTY, (char *)NULL) == -1) {
+        return posix_error();
+    }
+#else /* defined(HAVE_TTYNAME) */
+    /* Fallback method (archaic); from Advanced Programming in the UNIX(R)
+     * Environment, Third edition, 2013, Section 9.6 - Controlling Terminal:
+     * "Systems derived from UNIX System V allocate the controlling
+     * terminal for a session when the session leader opens the first
+     * terminal device that is not already associated with a session, as
+     * long as the call to open does not specify the O_NOCTTY flag." */
+    char *tmppath = ttyname(fd);
+    if (tmppath == NULL) {
+        return posix_error();
+    }
+
+#define CLOSE_IF_NOT_FD(otherfd) \
+    if (fd != otherfd) { \
+        close(otherfd); \
+    } \
+
+    CLOSE_IF_NOT_FD(0);
+    CLOSE_IF_NOT_FD(1);
+    CLOSE_IF_NOT_FD(2);
+
+    int tmpfd = open(tmppath, O_RDWR);
+    if (tmpfd == -1) {
+        return posix_error();
+    }
+    close(tmpfd);
+#endif /* TIOCSCTTY */
+
+    /* The tty becomes stdin/stdout/stderr */
+    if (dup2(fd, 0) == -1 || dup2(fd, 1) == -1 || dup2(fd, 2) == -1) {
+        return posix_error();
```

Comment:
aren't there potentially some successful dup2 calls to clean up from in this error case?

Reply:
Good catch. Suppose that `dup2(fd, 0)` succeeds and `dup2(fd, 1)` fails. The only cleanup that comes to mind is `close(0)`. However, by that time, we are already leaving the process in an unpleasant state since the original `0`, `1`, `2` are now closed because of the `CLOSE_IF_NOT_FD(...)` above. I feel like we should remove the error checks from the `dup2()` calls here if you think that is acceptable (yes/no); none of glibc, freebsd, netbsd, openbsd, cpython's `pty.py` check for `dup2()` errors here (see references below), except for glibc looping while there is failure with `errno == EBUSY` (race condition with `open()` and `dup()`). I now believe that there was no strong reason for me adding those `dup2()` error checks when I wrote this code earlier.

1. https://github.com/lattera/glibc/blob/master/login/login_tty.c
2. https://github.com/freebsd/freebsd-src/blob/main/lib/libutil/login_tty.c
3. https://github.com/openbsd/src/blob/master/lib/libutil/login_tty.c
4. https://github.com/NetBSD/src/blob/trunk/lib/libutil/login_tty.c
5. https://github.com/python/cpython/blob/main/Lib/pty.py

Reply:
I guess there's not much we can do in the way of total recovery to the original state here. I still like letting an error be reported though. So lets just leave this as is. In theory this isn't something that should really come up.

Reply:
I guess there's not much we can do in the way of total recovery to the original state here. I still like letting an error be reported though. So lets just leave this as is. In theory this isn't something that should really come up.